notebook/tests/launchnotebook.py

```diff
@@ -90,23 +91,41 @@ def request(cls, verb, path, **kwargs):
             url_path_join(cls.base_url(), path),
             **kwargs)
         return response
-
+    
     @classmethod
     def setup_class(cls):
-        cls.home_dir = TemporaryDirectory()
-        data_dir = TemporaryDirectory()
+        cls.tmp_dir = TemporaryDirectory()
+        def tmp(*parts):
+            path = os.path.join(cls.tmp_dir.name, *parts)
+            try:
+                os.makedirs(path)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise
+            return path
+        
+        cls.home_dir = tmp('home')
+        data_dir = cls.data_dir = tmp('data')
+        config_dir = cls.config_dir = tmp('config')
+        runtime_dir = cls.runtime_dir = tmp('runtime')
         cls.env_patch = patch.dict('os.environ', {
-            'HOME': cls.home_dir.name,
+            'HOME': cls.home_dir,
```

Comment:
Is it definitely necessary to fake HOME as well as all the other things? We do this in ipykernel, IIRC, and I find it annoying for testing locally with `--user` installs, because with HOME changed, it no longer finds my installed packages.